### PR TITLE
Fix export for gemma4 and add Integration tests

### PR DESCRIPTION
**Pull request by Cyrilvallez**
# What does this PR do?

As per the title

### CODE REVIEW

**Comment by Cyrilvallez** [comment]
run-slow: gemma4

**Comment by Cyrilvallez** [comment]
run-slow: gemma4

File: tests/models/gemma4/test_modeling_gemma4.py [hunk 1]
```diff
@@ -419,12 +413,12 @@ def test_generate_from_random_inputs_embeds(self):
...
 
 
-@unittest.skip("Integration Tests are not up-to-date yet! TODO Cyril: update me pretty pretty please!")
 @slow
 @require_torch_accelerator
 class Gemma4IntegrationTest(unittest.TestCase):
     def setUp(self):
-        self.processor = Gemma4Processor.from_pretrained("google/gemma-4-e2b-it", padding_side="left")
+        self.model_name = "google/gemma-4-E2B-it"
+        self.processor = Gemma4Processor.from_pretrained(self.model_name, padding_side="left")
```
Comment (vasqu) [inline tests/models/gemma4/test_modeling_gemma4.py]: Imo, padding side should be fixed on the hub, no?

File: tests/models/gemma4/test_modeling_gemma4.py
Comment (vasqu) [inline tests/models/gemma4/test_modeling_gemma4.py]: We should use `url_to_local_path` for caching tbh, seems like this is copy paste from the old gemmas which already do it wrong as well 🤔 https://github.com/huggingface/transformers/blob/dd6e8c5e25a3281c4e1cb34c270edfb0a8699a74/utils/fetch_hub_objects_for_ci.py#L11

Comment (vasqu) [inline tests/models/gemma4/test_modeling_gemma4.py]: Same here then (and maybe elsewhere)

File: tests/models/gemma4/test_modeling_gemma4.py [hunk 2]
```diff
@@ -680,187 +524,130 @@ def test_model_4b_multiimage(self):
...
+        out = model.generate(**inputs, max_new_tokens=20, do_sample=False, cache_implementation="static")
+        output_text = tokenizer.batch_decode(out[:, input_size:])
 
-        EXPECTED_COMPLETIONS = [" and I'm going to take a walk.\n\nI really enjoy the scenery, and I'", ", green, yellow, orange, purple, brown, black, white, gray.\n\nI'"]  # fmt: skip
-        self.assertEqual(output_text, EXPECTED_COMPLETIONS)
+        EXPECTED_COMPLETIONS_EAGER = Expectations(
+            {
+                ("cuda", 8): [
+                    "That sounds lovely! It seems like you're really enjoying the place you're in.\n\n",
+                    "Here are a few ways you could use or expand upon that list, depending on what you need:",
```
Comment (vasqu) [inline tests/models/gemma4/test_modeling_gemma4.py]: Imo, we may cut to just 16 tokens and have the same output then? Just seeing that it slightly differs at the end of the second output

File: tests/models/gemma4/test_modeling_gemma4.py [hunk 3]
```diff
@@ -680,187 +524,130 @@ def test_model_4b_multiimage(self):
...
-        self.assertEqual(model.config.cache_implementation, "hybrid")
+        model = Gemma4ForConditionalGeneration.from_pretrained(self.model_name)
+        tokenizer = AutoTokenizer.from_pretrained(self.model_name)
 
-        # Export + hybrid cache
-        model.eval()
         exportable_module = TorchExportableModuleForDecoderOnlyLM(model, batch_size=1, max_cache_len=1024)
         exported_program = exportable_module.export(
-            input_ids=torch.tensor([[1]], dtype=torch.long, device=model.device),
+            input_ids=torch.tensor([[1]], dtype=torch.long),
```
Comment (vasqu) [inline tests/models/gemma4/test_modeling_gemma4.py]: Intentional to run on cpu?

File: src/transformers/integrations/executorch.py [hunk 4]
```diff
@@ -702,9 +715,22 @@ def __init__(
...
+            ]
+            num_heads = [
+                config.num_global_key_value_heads
+                if layer == "full_attention" and config.attention_k_eq_v
+                else config.num_key_value_heads
+                for layer in config.layer_types[: -config.num_kv_shared_layers]
+            ]
+        else:
+            head_dim = getattr(config, "head_dim", config.hidden_size // config.num_attention_heads)
+            num_heads = getattr(config, "num_key_value_heads", config.num_attention_heads)
```
Comment (vasqu) [inline src/transformers/integrations/executorch.py]: Maybe a small helper function instead of repeating twice?

**Review by vasqu — APPROVED** [review]
Overall smaller things imo
- Local cache image
- Maybe hub update gemma4 for the padding side
